benches/bignumber_benchmark.rs

```diff
@@ -0,0 +1,147 @@
+use num_bigint::BigInt;
+use openssl::bn::{BigNum, BigNumContext, BigNumRef};
+use rug::{Complete, Integer};
+
+const PRIME_SIZE: usize = 1024;
+
+criterion_group! {
+    name = slow_primegen;
+    config = Criterion::default().sample_size(10);
+    targets = compare_prime_gen
+}
+criterion_group!(benches, compare_modpow);
+criterion_main!(benches, slow_primegen);
+
+fn compare_prime_gen(c: &mut Criterion) {
+    println!("⚠️ gnu prime generation benchmark will only be correct if the libpaillier library has feature \"gnu\" enabled!");
+    c.bench_function("openssl prime gen", |b| b.iter(ossl_prime_gen));
+    c.bench_function("bigint prime gen", |b| b.iter(bigint_prime_gen));
+    c.bench_function("gnu prime gen", |b| b.iter(gnu_prime_gen));
+}
+
+/// Use the openssl wrapper directly.
+fn ossl_prime_gen() {
+    let mut prime = BigNum::new().unwrap();
+    prime
+        .generate_prime(PRIME_SIZE as i32, true, None, None)
+        .unwrap();
+}
+
+/// The Rust num_bigint crate doesn't have safe-prime generation methods directly. Instead,
+/// use the glass pumpkin crate to create a `BigUint`.
+fn bigint_prime_gen() {
+    let mut rng = rand::thread_rng();
+    safe_prime::from_rng(PRIME_SIZE, &mut rng).unwrap();
+}
+
+/// The GNU prime generation method is hard to use, so assume that the BigNumber crate is configured
+/// to use gnu.
+fn gnu_prime_gen() {
+    BigNumber::safe_prime(PRIME_SIZE);
+}
+
+/// Modular exponentiation is a common operation in the protocol. A typical usage is in Paillier
+/// encryption. With our parameter settings, this would be a `b^a mod M`, where:
+/// - b: the base is either from Z*_N or is 1+N, where N is a 2048-bit modulus.
+/// - a: the exponent is either a message in Z_N or N itself (so, 2048 bits).
+/// - M: the modulus is the square of the 2048-bit modulus N
+///
+/// Setup allows us to not count the time to initiate variables in the benchmarking.
+fn compare_modpow(c: &mut Criterion) {
+    let mut base_rng = rand::thread_rng();
+    let mut exp_rng = rand::thread_rng();
+
+    // Get a hex-encoded element in Z*_N, where N is the output of `modulus()`.
+    let mut base = || -> String {
+        let n = BigNumber::from_slice(hex::decode(modulus_string()).unwrap());
+        let z_star: BigNumber = BigNumber::from_rng(&(n - 1), &mut base_rng) + 1;
+        z_star.to_bytes().encode_hex()
+    };
+
+    // Get a hex-encoded element in Z_N, where N is the output of `modulus()`.
+    let mut exponent = || -> String {
+        let n = BigNumber::from_slice(hex::decode(modulus_string()).unwrap());
+        BigNumber::from_rng(&n, &mut exp_rng)
+            .to_bytes()
+            .encode_hex()
+    };
+
+    let modulus = &to_ossl(modulus_string()) * &to_ossl(modulus_string());
+    c.bench_function("openssl modpow", |b| {
+        b.iter_batched(
+            || (to_ossl(base()), to_ossl(exponent())),
+            |(base, exp)| ossl_modpow(base, exp, &modulus),
+            BatchSize::SmallInput,
```

Comment:
It took my eyes a bit to parse the closures!

Reply:
Yeah, I don't love the volume of tiny functions and closures in this benchmark :| But it'll do!